[16.0][ADD] l10n_es_verifactu_oca: Comunicación VERI*FACTU

zamberjo opened the pull request:
Movemos a este PR el nuevo módulo `l10n_es_aeat_verifactu`

Depends on:

- [x]  [[16.0][IMP] Refactor para futura implementación Veri*FACTU.](https://github.com/OCA/l10n-spain/pull/3572)

Commit:
[ADD] l10n_es_verifactu_oca: Comunicación VERI*FACTU

Envío de los registros de facturación a la sede electrónica de la
Agencia Tributaria en el momento de su producción

AndreuOForgeFlow committed:
[IMP] l10n_es_aeat_verifactu_oca: Activar Journal Hash para compañía

almumu commented:
Añadido la lógica que generará el hash de las facturas. Basado en la documentación de la aeat
https://www.agenciatributaria.es/static_files/AEAT_Desarrolladores/EEDD/IVA/VERI-FACTU/Veri-Factu_especificaciones_huella_hash_registros.pdf
De momento son campos compute y sin store True para poder ir avanzando.

- [ ] Pendiente decidir de qué modo se va a saber cuál es la factura anterior enviada, ya que el hash de una factura a su vez está compuesto por el hash de la anterior factura enviada.

- [ ] Pendiente decidir en qué momento generaremos el hash y que se guarde en la factura.

- [ ] El tipo de documento de verifactu, de momento es un selection, si esos son los tipos definitivos podríamos convertirlo a un modelo y un data

pedrobaeza commented:
Se puede ya hacer rebase.

zamberjo commented:
> Se puede ya hacer rebase.

Hecho y he añadido los últimos cambios para setear el campo `aeat_sending_enabled`

almumu committed:
[IMP] l10n_es_verifactu_oca: Add verifactu hash code + send to AEAT

giarve commented on l10n_es_aeat_verifactu/models/account_move.py:
```diff
@@ -65,3 +103,52 @@ def _get_document_serial_number(self):
+        return self.amount_total
+
+    def _get_verifactu_previous_hash(self):
+        # TODO store it? search it by some kind of sequence?
```
Indeed Odoo chose to create a sequence when creating the journal and uses the sequence to find the previous entry.

https://github.com/odoo/odoo/blob/d8a20717c830a7cddaad4889bec182f0fa5f786a/addons/account/models/account_move.py#L2548-L2551

https://github.com/odoo/odoo/blob/d8a20717c830a7cddaad4889bec182f0fa5f786a/addons/account/models/account_journal.py#L950

anmarmo1 replied:
Yes, but the problem in this case is that the previous record is not the previous one from the same journal. A possible solution is to implement a unique sequence for all invoices, but we also need to know if they have been sent and received by the AEAT.

giarve replied:
Lo que se me ocurre es una especie de cola, así se elimina el problema de la concurrencia (y se puede tener tantas secuencias como diarios). La cola FIFO guarda la referencia o hash de la última factura en un campo global por orden de inserción.

Faltaría decidir cuando meter la factura en la cola que generará el hash y la enviará a AEAT. No sé si por la ley hay obligatoriedad de hacerlo cuando se le da a enviar factura o solamente al guardar, o pasa como el SII que puede ser manual. En cualquier caso sería meterlo en la cola desde cualquier diario y encadenar facturas de diferentes diarios, validadas por diferentes usuarios, o demás **concurrencias**, gracias a la cola no supondrían un problema.

es por dar ideas.

pedrobaeza replied:
Me resulta muy extraño que esa factura anterior tenga que ser global en lugar de por diario. La AEAT es muy consciente de las cuestiones técnicas asociadas a tener más de una secuencia de numeración (como múltiples canales de venta).

ljsalvatierra-factorlibre replied:
Según mi interpretación de la documentación, el encadenamiento es por _SIF_ (Sistema Informático de Facturación).

![imagen](https://github.com/user-attachments/assets/5b1cb2ad-8588-4cd9-b796-97ecbdfaae34)

```
Caso 2: registro de facturación –en este caso de alta–
con registro de facturación anterior existente en el SIF
(siendo, por tanto, el segundo registro o sucesivo)
```

Lo interpreto de ese modo porque dice `factura anterior existente en el SIF`.

En caso de ser mi interpretación correcta, entiendo que habría que crear un SIF para la facturación de Odoo y uno por TPV, si no recuerdo mal así es como funciona en TicketBAI.

anmarmo1 replied:
Si por eso interpretamos que el encadenamiento era con el anterior enviado y eso excluye el orden por diario pero realmente tampoco habla de series en ningun lado que yo haya visto...

SoniaViciana commented:
Hola @zamberjo @almumu 

Veo que falta la parte del hash:

![image](https://github.com/user-attachments/assets/4f642a04-2405-4063-af9e-559c3d1848e4)

La documentación técnica de la AEAT está disponible (en "Borrador"), y en cuanto al tema de controlar la factura anterior enviada a Verifactu, si no me equivoco, es algo que ya requiere y está disponible dentro de TicketBai, así que seguramente lo podremos aprovechar.

¿Os parece bien si trabajamos de nuestro lado en la parte del Hash? ¿Veis necesario reunirnos antes? lo comento por no trabajar en paralelo en lo mismo y que nos repartamos la carga de trabajo.

Quedo a la espera de vuestros comentarios.
Gracias!

-------
Fuente: https://www.agenciatributaria.es/AEAT.desarrolladores/Desarrolladores/_menu_/Documentacion/IVA/Sistemas_Informaticos_de_Facturacion_y_Sistemas_VERI_FACTU/Sistemas_Informaticos_de_Facturacion_y_Sistemas_VERI_FACTU.html

anmarmo1 commented:
> Hola @zamberjo @almumu
> 
> Veo que falta la parte del hash:
> 
> ![image](https://private-user-images.githubusercontent.com/46522913/372090427-4f642a04-2405-4063-af9e-559c3d1848e4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc3NjYyNDMsIm5iZiI6MTcyNzc2NTk0MywicGF0aCI6Ii80NjUyMjkxMy8zNzIwOTA0MjctNGY2NDJhMDQtMjQwNS00MDYzLWFmOWUtNTU5YzNkMTg0OGU0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDAxVDA2NTkwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVmZTM2NjA5MmFhOTFmZDI3NWMwYjU0NjgzYWMwYTU3N2EwZTYzZGE4Y2JhMmZlMzQ0ZDJiNjA3NDI3NDgxN2EmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Me8K3_xpw-v9N9tkE70X17cQ05pnP5osNBdfSLzFmrc)
> 
> La documentación técnica de la AEAT está disponible (en "Borrador"), y en cuanto al tema de controlar la factura anterior enviada a Verifactu, si no me equivoco, es algo que ya requiere y está disponible dentro de TicketBai, así que seguramente lo podremos aprovechar.
> 
> ¿Os parece bien si trabajamos de nuestro lado en la parte del Hash? ¿Veis necesario reunirnos antes? lo comento por no trabajar en paralelo en lo mismo y que nos repartamos la carga de trabajo.
> 
> Quedo a la espera de vuestros comentarios. Gracias!
> 
> Fuente: https://www.agenciatributaria.es/AEAT.desarrolladores/Desarrolladores/_menu_/Documentacion/IVA/Sistemas_Informaticos_de_Facturacion_y_Sistemas_VERI_FACTU/Sistemas_Informaticos_de_Facturacion_y_Sistemas_VERI_FACTU.html

Hola Sonia
Perfecto, si quereis encargaros vosotros de la parte del HASH no hay problema, si quereis una reunión tampoco hay problema.
El tema del hash es que, el que implementa Odoo no vale, por que los valores para construirlo para el verifactu son distintos y además la complejidad reside en saber cual es el último registro que ha sido enviado a la AEAT por el tema del encadenamiento. Con respecto a utilizar la parte de ticketbai para esto, podéis revisar como se implemento como inspiración, pero no vale para la implementación del verifactu ya que esta implementado con una librería de python externa y en los OCA Days los PSC nos indicaron que de esta manera no se fusionaría. Por nuestro lado, si queréis podemos empezar con las llamadas al wsdl de la AEAT y así avanzamos.

ljsalvatierra-factorlibre commented:
> > Hola @zamberjo @almumu
> > Veo que falta la parte del hash:
> > ![image](https://private-user-images.githubusercontent.com/46522913/372090427-4f642a04-2405-4063-af9e-559c3d1848e4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc3NjYyNDMsIm5iZiI6MTcyNzc2NTk0MywicGF0aCI6Ii80NjUyMjkxMy8zNzIwOTA0MjctNGY2NDJhMDQtMjQwNS00MDYzLWFmOWUtNTU5YzNkMTg0OGU0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDAxVDA2NTkwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVmZTM2NjA5MmFhOTFmZDI3NWMwYjU0NjgzYWMwYTU3N2EwZTYzZGE4Y2JhMmZlMzQ0ZDJiNjA3NDI3NDgxN2EmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Me8K3_xpw-v9N9tkE70X17cQ05pnP5osNBdfSLzFmrc)
> > La documentación técnica de la AEAT está disponible (en "Borrador"), y en cuanto al tema de controlar la factura anterior enviada a Verifactu, si no me equivoco, es algo que ya requiere y está disponible dentro de TicketBai, así que seguramente lo podremos aprovechar.
> > ¿Os parece bien si trabajamos de nuestro lado en la parte del Hash? ¿Veis necesario reunirnos antes? lo comento por no trabajar en paralelo en lo mismo y que nos repartamos la carga de trabajo.
> > Quedo a la espera de vuestros comentarios. Gracias!
> > Fuente: https://www.agenciatributaria.es/AEAT.desarrolladores/Desarrolladores/_menu_/Documentacion/IVA/Sistemas_Informaticos_de_Facturacion_y_Sistemas_VERI_FACTU/Sistemas_Informaticos_de_Facturacion_y_Sistemas_VERI_FACTU.html
> 
> Hola Sonia Perfecto, si quereis encargaros vosotros de la parte del HASH no hay problema, si quereis una reunión tampoco hay problema. El tema del hash es que, el que implementa Odoo no vale, por que los valores para construirlo para el verifactu son distintos y además la complejidad reside en saber cual es el último registro que ha sido enviado a la AEAT por el tema del encadenamiento. Con respecto a utilizar la parte de ticketbai para esto, podéis revisar como se implemento como inspiración, pero no vale para la implementación del verifactu ya que esta implementado con una librería de python externa y en los OCA Days los PSC nos indicaron que de esta manera no se fusionaría. Por nuestro lado, si queréis podemos empezar con las llamadas al wsdl de la AEAT y así avanzamos.

Hola @anmarmo1, ¿con librería de python externa te refieres a la librería de Javascript `tbai.js`?

anmarmo1 commented:
> > > Hola @zamberjo @almumu
> > > Veo que falta la parte del hash:
> > > ![image](https://private-user-images.githubusercontent.com/46522913/372090427-4f642a04-2405-4063-af9e-559c3d1848e4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc3NjYyNDMsIm5iZiI6MTcyNzc2NTk0MywicGF0aCI6Ii80NjUyMjkxMy8zNzIwOTA0MjctNGY2NDJhMDQtMjQwNS00MDYzLWFmOWUtNTU5YzNkMTg0OGU0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDAxVDA2NTkwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVmZTM2NjA5MmFhOTFmZDI3NWMwYjU0NjgzYWMwYTU3N2EwZTYzZGE4Y2JhMmZlMzQ0ZDJiNjA3NDI3NDgxN2EmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Me8K3_xpw-v9N9tkE70X17cQ05pnP5osNBdfSLzFmrc)
> > > La documentación técnica de la AEAT está disponible (en "Borrador"), y en cuanto al tema de controlar la factura anterior enviada a Verifactu, si no me equivoco, es algo que ya requiere y está disponible dentro de TicketBai, así que seguramente lo podremos aprovechar.
> > > ¿Os parece bien si trabajamos de nuestro lado en la parte del Hash? ¿Veis necesario reunirnos antes? lo comento por no trabajar en paralelo en lo mismo y que nos repartamos la carga de trabajo.
> > > Quedo a la espera de vuestros comentarios. Gracias!
> > > Fuente: https://www.agenciatributaria.es/AEAT.desarrolladores/Desarrolladores/_menu_/Documentacion/IVA/Sistemas_Informaticos_de_Facturacion_y_Sistemas_VERI_FACTU/Sistemas_Informaticos_de_Facturacion_y_Sistemas_VERI_FACTU.html
> > 
> > 
> > Hola Sonia Perfecto, si quereis encargaros vosotros de la parte del HASH no hay problema, si quereis una reunión tampoco hay problema. El tema del hash es que, el que implementa Odoo no vale, por que los valores para construirlo para el verifactu son distintos y además la complejidad reside en saber cual es el último registro que ha sido enviado a la AEAT por el tema del encadenamiento. Con respecto a utilizar la parte de ticketbai para esto, podéis revisar como se implemento como inspiración, pero no vale para la implementación del verifactu ya que esta implementado con una librería de python externa y en los OCA Days los PSC nos indicaron que de esta manera no se fusionaría. Por nuestro lado, si queréis podemos empezar con las llamadas al wsdl de la AEAT y así avanzamos.
> 
> Hola @anmarmo1, ¿con librería de python externa te refieres a la librería de Javascript `tbai.js`?

Hola @ljsalvatierra-factorlibre realmente me referia a este PR https://github.com/OCA/l10n-spain/pull/3547

almumu commented:
Hemos hecho una primera versión funcional de envío de facturas a verifactu.

El desarrollo está hecho de la forma más parecida posible al módulo del sii, para que en caso de que haya que unificar o modificar procesos, sea los más entendible posible para todos los que toquemos código.

Falta mucho por hacer pero ya es un paso, cualquier aportación es bienvenida.

Para poder probarlo:

- Tener certificado
- Activar verifactu en la compañía y establecer que es entorno de pruebas.
- Configurar en la posición fiscal la clave de impuesto (Verifactu Tax Key): IVA y clave de registro verifactu: Operación de Régimen General, para que al crear la factura se rellenen los datos.
- Hay un mapeo de impuestos como en el sii, algunos están ya rellenados pero faltará revisarlo.
- De momento el envío se hace desde el botón manual en la factura, una vez validada. Todavía no está implementado el envío con queue job.
- Sólo se han hecho pruebas de envío con facturas normales de régimen nacional o recargo de equivalencia sin impuestos especiales. Como la parte del encadenamiento con el registro anterior en el hash sigue sin estar clara, de momento se envía siempre como si fuera el primer registro para poder enviarlo.
- No están contemplados aún casos como facturas simplificadas, terceros, no sujetas.., tampoco modificaciones y anulaciones de facturas.
- Los datos del desarrollador: IdSistemaInformatico, NumeroInstalacion, .. nos los hemos inventado un poco, esto habrá que verlo.

syci commented:
Hola, por mi parte empezaré a revisar/probar y demás para poder desarrollar la parte necesaria para la ATC en Canarias.

ozono commented:
He enviado un PR https://github.com/aurestic/l10n-spain/pull/62 a los compañeros de Aurestic como punto de partida para el QR en las facturas. Es una implementación funcional, aunque hay detalles por discutir y revisar. Aprovecho para dar las gracias a todos por el trabajo realizado.

syci commented:
@pedrobaeza veo que en los mapeos los impuestos del IGIC no están en https://github.com/aurestic/l10n-spain/tree/ADD/l10n_es_aeat_verifactu/l10n_es_aeat_verifactu/data , ¿Dónde se implementaría en el módulo del igic, en uno nuevo verifactu_igic ...... ?

pedrobaeza commented:
Eso es, @syci. Habrá que hacer un módulo puente.

ozono committed:
[IMP] l10n_es_verifactu_oca: Primera version QR VERI*FACTU

ozono committed:
[ADD] Primera version QR Verifactu (#62)

anmarmo1 commented:
@SoniaViciana @ljsalvatierra-factorlibre ¿Vais a implementar la parte del Encadenamiento? Es por seguir nosotros. 
El registro ha de ser único para todas las series.
![image](https://github.com/user-attachments/assets/34c5fcc4-f098-4a6b-88fc-df1a7ef7ef62)
Lo que no se que pasaría si el sistema esta caído, siempre hablan de "por orden cronológico de generación de registros de facturación en el SIF", he enviado la consulta a la AEAT por que no lo he visto en las FAQ

ljsalvatierra-factorlibre commented:
> @SoniaViciana @ljsalvatierra-factorlibre ¿Vais a implementar la parte del Encadenamiento? Es por seguir nosotros. El registro ha de ser único para todas las series. ![image](https://private-user-images.githubusercontent.com/15904368/405925371-34c5fcc4-f098-4a6b-88fc-df1a7ef7ef62.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc2MTk3MjMsIm5iZiI6MTczNzYxOTQyMywicGF0aCI6Ii8xNTkwNDM2OC80MDU5MjUzNzEtMzRjNWZjYzQtZjA5OC00YTZiLTg4ZmMtZGYxYTdlZjdlZjYyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTIzVDA4MDM0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE0ODA4MmRhMzQ2NGYzY2IwOTNjZTk1OWVkZTFlYzIxZWMyMjdhYmYyZDQ4MTllZjFiMWZhZmQ5NTg0Njc3YjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.4F8V8Gw9wEJxU2cbg2Dr69ecd53msMBCcZirTvLkkw4) Lo que no se que pasaría si el sistema esta caído, siempre hablan de "por orden cronológico de generación de registros de facturación en el SIF", he enviado la consulta a la AEAT por que no lo he visto en las FAQ

Hola @anmarmo1 estamos realizando el desarrollo para tener un SIF para el backend de Odoo y un SIF por cada TPV (`pos.config`), del mismo modo que TicketBAI.

anmarmo1 commented:
@ljsalvatierra-factorlibre ok entonces esperamos. ¡Gracias!

ljsalvatierra-factorlibre commented:
> @ljsalvatierra-factorlibre ok entonces esperamos. ¡Gracias!

Podéis revisar los cambios aquí https://github.com/OCA/l10n-spain/pull/3973
Me decís por favor en ese PR cualquier cambio que veáis necesario, gracias.
Cuando esté todo listo vemos cómo integrarlo.

ljsalvatierra-factorlibre commented:
> > @ljsalvatierra-factorlibre ok entonces esperamos. ¡Gracias!
> 
> Podéis revisar los cambios aquí #3973 Me decís por favor en ese PR cualquier cambio que veáis necesario, gracias. Cuando esté todo listo vemos cómo integrarlo.

Estoy ahora con el desarrollo de `l10n_es_aeat_verifactu_pos`.

anmarmo1 commented:
> > @ljsalvatierra-factorlibre ok entonces esperamos. ¡Gracias!
> 
> Podéis revisar los cambios aquí #3973 Me decís por favor en ese PR cualquier cambio que veáis necesario, gracias. Cuando esté todo listo vemos cómo integrarlo.

Ok! Revisamos

almumu commented:
Muchas gracias por vuestra aportación @ljsalvatierra-factorlibre  , en el pr https://github.com/OCA/l10n-spain/pull/3973
a nivel funcional lo hemos probado y está correcto, y la idea de guardarlo a nivel de compañía es buena.

Aunque la parte de anulación de facturas aún no está hecha, ahí también se envía el encadenamiento, habrá que tener cuidado a futuro de no invocar a _get_chaining_invoice_dict ya que se machacaría el valor del encadenamiento. O bien añadir una comprobación de que si ya está establecido el verifactu_last_invoice_id en la factura, no lo modifique ni en la factura ni en la compañía.

Aun así siguen habiendo varias dudas que aún no sabemos como se van a implementar y que tendremos que llegar a un consenso entre todos...
En qué momento se enviarán las facturas? al validarlas? generamos una cola de envío? porque si hay múltiples usuarios validando y enviando facturas, puede haber problemas de concurrencia con lo del encadenamiento. O por ejemplo si se va la luz durante un envío, y no se sabe si se ha enviado a verifactu porque no se había obtenido la respuesta de la aeat de si el envío ha sido correcto.., y ese tipo de cosas que pueden afectar al encadenamiento.

Tampoco tenemos claro cuál es la fecha-hora que deberíamos enviar en la huella (FechaHoraHusoGenRegistro), ahora mismo es la de creación ([_get_verifactu_registration_date](https://github.com/OCA/l10n-spain/blob/477fd4fbf8f13185a8a7a7f1bdb3ab0c86bb9cbd/l10n_es_aeat_verifactu/models/account_move.py#L133)), pero eso no es viable porque si pasan más de 120 segundos de esa fecha-hora de creación hasta que se envía a verifactu (lo cual es  muy probable que pase), la aeat la rechaza con el error "2004 El valor del campo FechaHoraHusoGenRegistro debe ser la fecha actual del sistema de la AEAT, admitiéndose un margen de error de: 120 segundos". Quizá lo correcto sería crear un nuevo campo de fecha envío verifactu que sea cuando realmente se envía a verifactu y que se guarde en la factura y sea ese el que se use para obtener el hash.


En cualquier caso, de momento podéis hacernos un pr a nuestra rama con estos cambios https://github.com/OCA/l10n-spain/pull/3973/commits/03f46ef125f3964b4d7da41c238950fdcbde9222 y así los mergeamos mientras se decide todo lo demás.
Agradeceros de nuevo vuestra contribución
Saludos

ljsalvatierra-factorlibre commented:
> Muchas gracias por vuestra aportación @ljsalvatierra-factorlibre , en el pr #3973 a nivel funcional lo hemos probado y está correcto, y la idea de guardarlo a nivel de compañía es buena.
> 
> Aunque la parte de anulación de facturas aún no está hecha, ahí también se envía el encadenamiento, habrá que tener cuidado a futuro de no invocar a _get_chaining_invoice_dict ya que se machacaría el valor del encadenamiento. O bien añadir una comprobación de que si ya está establecido el verifactu_last_invoice_id en la factura, no lo modifique ni en la factura ni en la compañía.
> 
> Aun así siguen habiendo varias dudas que aún no sabemos como se van a implementar y que tendremos que llegar a un consenso entre todos... En qué momento se enviarán las facturas? al validarlas? generamos una cola de envío? porque si hay múltiples usuarios validando y enviando facturas, puede haber problemas de concurrencia con lo del encadenamiento. O por ejemplo si se va la luz durante un envío, y no se sabe si se ha enviado a verifactu porque no se había obtenido la respuesta de la aeat de si el envío ha sido correcto.., y ese tipo de cosas que pueden afectar al encadenamiento.
> 
> Tampoco tenemos claro cuál es la fecha-hora que deberíamos enviar en la huella (FechaHoraHusoGenRegistro), ahora mismo es la de creación ([_get_verifactu_registration_date](https://github.com/OCA/l10n-spain/blob/477fd4fbf8f13185a8a7a7f1bdb3ab0c86bb9cbd/l10n_es_aeat_verifactu/models/account_move.py#L133)), pero eso no es viable porque si pasan más de 120 segundos de esa fecha-hora de creación hasta que se envía a verifactu (lo cual es muy probable que pase), la aeat la rechaza con el error "2004 El valor del campo FechaHoraHusoGenRegistro debe ser la fecha actual del sistema de la AEAT, admitiéndose un margen de error de: 120 segundos". Quizá lo correcto sería crear un nuevo campo de fecha envío verifactu que sea cuando realmente se envía a verifactu y que se guarde en la factura y sea ese el que se use para obtener el hash.
> 
> En cualquier caso, de momento podéis hacernos un pr a nuestra rama con estos cambios [03f46ef](https://github.com/OCA/l10n-spain/commit/03f46ef125f3964b4d7da41c238950fdcbde9222) y así los mergeamos mientras se decide todo lo demás. Agradeceros de nuevo vuestra contribución Saludos

Hola @almumu, muchas gracias por revisarlo, sobre el envío:

```
"El tema de encadenamiento de la huella de las facturas esa huella su encadenamiento se genera en funcion de la serie de 
facturacion, es decir hago la A/1, A/2 se encadena con A/1 pero si ahora hago la C/1 ¿se encadena con A/2 o realmente no 
porque es la primera de la serie C?",Ramon Sanchez,'-,,10/28/2024 13:56:15,,
```
```
El encadenamiento se realiza por orden cronológico de generación de registros de facturación, independientemente de si es un 
registro de alta o anulación o del nº de factura o serie del mismo.  No está relacionado con el orden en que los registros son 
comunicados (pueden desordenarse en la comunicación si hubiera que hacer reenvíos), ni con la serie/número de la factura 
(aunque normalmente suela seguir una cierta pauta temporal), sino con el orden en que son generados.  Cualquier registro de 
facturación nuevo tiene que ir encadenado al registro de facturación inmediatamente anterior. Eso quiere decir que el bloque 
"Encadenamiento" se rellenará con los datos del bloque "RegistroAnterior" del registro de facturación anterior y el campo 
“Huella”.
```

aritzolea commented on l10n_es_aeat_verifactu/models/account_journal.py:
```diff
@@ -0,0 +1,74 @@
+                and record.company_id.country_code == "ES"
+                and record.type == "sale"
+            ):
+                record.restrict_mode_hash_table = True
+                record.restrict_mode_hash_table_readonly = True
```
He detectado que esto no está llamando al write de `account.journal`, y está dando problemas al activar Verifactu en una compañía. Podríais actualizar esos dos campos mediante un write?

almumu replied:
hola @aritzolea , no sé muy bien qué problemas os estará dando esta parte pero si tienes el caso controlado, podéis hacer cualquier modificación necesaria en vuestra rama y añadirlo al pr que nos hagáis. Quizá sería buena idea que nos hagáis el pr a esta rama para poder mergearlo cuanto antes y que todos hagamos las pruebas y aportaciones sobre el mismo código, como hicimos con el de @ozono , y seguir avanzando.

aritzolea replied:
Buenas @almumu
Estoy con pruebas en mi local, intento abriros el PR entre hoy y mañana.

almumu replied:
genial! muchas gracias @aritzolea

aritzolea replied:
Buenas! Os he abierto PR

aritzolea replied:
He abierto otro PR para corregir pre-commit y tests

almumu replied:
> He abierto otro PR para corregir pre-commit y tests

merged. He hecho otro commit porque todavía daba error el precommit

aritzolea commented on l10n_es_aeat_verifactu/models/account_journal.py:
```diff
@@ -0,0 +1,74 @@
+    # no vamos autilizar el hash de odoo porque la estructura no nos sirve
+    # para verifactu, por lo que este código no nos terminaría de valer.
+    # De momento lo dejamos hasta saber cómo vamos a controlar
+    # el tema de la factura anterior enviada a verifactu para el cálculo
```
No entiendo del todo por que se ha hecho esto de esta forma. pero sigue siendo necesario después de la implementación del encadenamiento en https://github.com/OCA/l10n-spain/pull/3973?

Sí que habrá que controlar la modificación de las facturas ya enviadas, pero la primera parte no me queda clara.

almumu replied:
se hizo en su momento para utilizar la parte de odoo que mediante el restrict_mode_hash_table del account.move (que es un related al journal) controla la modificación de facturas, no permite pasarlas a borrador, controla qué campos se pueden modificar, etc. Lo que no se usa es ese ese propio hash para enviarlo a verifactu como hash puesto que no tiene nada que ver el formato de ese hash con el que pide verifactu.

aritzolea committed:
[IMP] l10n_es_aeat_verifactu: Fixes on views, hash calculated on invoice post and refactor + tests

aritzolea committed:
[IMP] l10n_es_aeat_verifactu_oca: Fixes on views, hash calculated on invoice post and refactor + tests

Fernando-Draw commented:
Hola!!, no se si es el sitio correcto para preguntar esto, pero dado que soy usuario de Odoo (17 Community) estaba buscando información sobre módulos de VERIFACTU aquí en OCA. Me he econtrado con este desarrollo y quisiera saber si de cara a futuro esta planteado su correspondiente PORT para las versiones de Odoo 17 y 18.

Gracias y pido perdón si no es el lugar correcto para preguntar esto, pero no soy muy bueno aun con github.

pedrobaeza commented:
Sí, cuando se termine la fase de desarrollo en esta versión, se llevará a todas las versiones superiores.

Fernando-Draw commented:
> Sí, cuando se termine la fase de desarrollo en esta versión, se llevará a todas las versiones superiores.

OK!! muchas gracias!! si necesitáis ayuda en cuanto a nivel de diseño gráfico (icono, o instrucciones o similar) si queréis puedo ayudar. En cuanto a programar no digo nada por que aun sigo aprendiendo.

almumu committed:
[IMP] l10n_es_verifactu_oca: Renaming + improvements + fixes

- Modificación de algunos métodos: control de encadenamiento, registro
  anterior y siguiente.
- Envío por cron.triggers
- Renombrado módulo para quitar lo de 'aeat', y también algunos modelos.
- Control de excepciones al hacer el post de la factura para que
  compruebe que la configuración es correcta.
- Algunas mejoras en el código.

almumu commented:
Buenos días! he subido cambios en cuanto al encadenamiento y la forma de enviar las facturas, siempre basándome en el módulo del sii para que todos sea lo más similar posible en cuanto a funcionalidad y entendible para tod@s en cuanto a código. Tambíén, tal y como hablamos en la reunion que hicimos de verifactu por discord, y dado que la AEAT obliga a que sea fácilmente consultable en caso de auditoría el registro anterior y siguiente de cada documento enviado,  el campo del registro anterior y siguiente se almacena en cada documento y se ha hecho como un campo Reference, de forma que si se quisiera utilizar el pos para verifactu, se podría perfectamente encadenar una factura con un pos order y viceversa. 
Cualquier cosa lo vamos hablando, saludos

IJOL commented:
Hola a todos:

Esta este PR usable? se puede ayudar?, que os falta? que necesitais?  y este es el código "oficial" de odoo para verifactu?

pedrobaeza commented:
Este código es el oficial de OCA, no de Odoo. En las [jornadas de Odoo de Cartagena](https://www.aeodoo.org/event/xviii-jornadas-odoo-2025-130/page/introduccion-xviii-jornadas-odoo-2025) y en los [Spanish OCA days](https://www.aeodoo.org/event/spanish-oca-days-2025-131/page/introduccion-spanish-oca-days-2025) se comentará extendidamente sobre esto.

IJOL commented:
No me instala, y no entiendo nada me faltan templates de iva en concreto account_tax_template_s_iva2b y algunos mas por lo que he podido comprobar, en ese momento me han temblado las piernas, seguro que algo sabeis que yo ignoro ;), agradeceria alguna indicacion que me haga salir del lio, gracias por adelantado

IJOL commented:
> Este código es el oficial de OCA, no de Odoo.

Lo tengo claro, pero eso tb nos hace la fuente oficial para odoo tb, pero si me exprese mal

pedrobaeza commented:
Sin mensajes concretos de error, no te podemos ayudar mucho más, pero empieza por tener un Odoo 100% actualizado  con todos sus módulos.

IJOL commented:
> Sin mensajes concretos de error, no te podemos ayudar mucho más, pero empieza por tener un Odoo 100% actualizado con todos sus módulos.

estaba claro,que no podia ser otra cosa que mi propia estulticia con las actualizaciones, ya esta, gracias

pedrobaeza commented:
De todas formas, empezar a mandar a Veri*Factu es obligatorio el 1 de enero de 2026, no el 1 de julio de 2025.

syci commented:
> Eso es, @syci. Habrá que hacer un módulo puente.

Hola @pedrobaeza y @Christian-RB hay tres nombres del nuevo módulo para la localización Canaria que podría encajar:
- l10n_es_atc_verifactu
- l10n_es_aeat_verifactu_igic
- l10n_es_aeat_verifactu_atc

¿Qué opináis?

pedrobaeza commented:
El primero me gusta más, pero también podemos plantearnos renombrar este módulo base a `l10n_es_verifactu`. No va a chocar con el de Odoo core, que se va a llamar `l10n_es_edi_verifactu`, y podría ser más adecuado, ya que no está solo vinculado a la AEAT. Así, la extensión podría ser `l10n_es_verifactu_atc` en su lugar.

syci commented:
Me parece adecuado lo que propones Pedro. Yo me quedo con l10n_es_verifactu_atc
________________________________
De: Pedro M. Baeza ***@***.***>
Enviado: viernes, mayo 23, 2025 12:12 PM
Para: OCA/l10n-spain ***@***.***>
Cc: Mario Montes ***@***.***>; Mention ***@***.***>
Asunto: Re: [OCA/l10n-spain] [16.0][ADD] l10n_es_aeat_verifactu (PR #3579)

[https://avatars.githubusercontent.com/u/7165771?s=20&v=4]pedrobaeza left a comment (OCA/l10n-spain#3579)<https://github.com/OCA/l10n-spain/pull/3579#issuecomment-2904092219>

El primero me gusta más, pero también podemos plantearnos renombrar este módulo base a l10n_es_verifactu. No va a chocar con el de Odoo core, que se va a llamar l10n_es_edi_verifactu, y podría ser más adecuado, ya que no está solo vinculado a la AEAT. Así, la extensión podría ser l10n_es_verifactu_atc en su lugar.

—
Reply to this email directly, view it on GitHub<https://github.com/OCA/l10n-spain/pull/3579#issuecomment-2904092219>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABWA3MSAF2224A4CV73MYC32737DNAVCNFSM6AAAAABHQAP67GVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSMBUGA4TEMRRHE>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

flachicasidoo committed:
[IMP] l10n_es_verifactu_oca: README + journal enable flag

flachicasidoo committed:
[IMP] l10n_es_verifactu: README + journal enable flag

ProcessControl-JLR committed:
[IMP+FIX] l10n_es_verifactu: Registration date + start date + search

- Añade la configuración de claves de registro y claves de impuestos a las plantillas de posiciones fiscales de la localización
- Añade la fecha de inicio para habilitar verifactu en las facturas.
- Corrige la búsqueda de empresas españolas y añade rollback en caso de
  error al enviar documentos.

ProcessControl-JLR committed:
[IMP+FIX] l10n_es_verifactu_oca: Registration date + start date + search

- Añade la configuración de claves de registro y claves de impuestos a las plantillas de posiciones fiscales de la localización
- Añade la fecha de inicio para habilitar verifactu en las facturas.
- Corrige la búsqueda de empresas españolas y añade rollback en caso de
  error al enviar documentos.

almumu committed:
[FIX+IMP] l10n_es_verifactu_oca: Taxes + general improvements

- Añadido algunos mapeos de impuestos que faltaban.
- Controlado que no se puedan validar facturas de venta si no están
  todos sus impuestos mapeados.
- Arreglado algunos errores de configuración y funcionamiento.
- Los impuestos no incluidos en el total no se tenían en cuenta y se
  enviaba un importe erróneo.
- El total impuestos y total factura en verifactu aparecen tanto en
  detalle que se envía, como en el hash como en el qr que se imprime y
  la url de verificación de factura, arreglado en todos los procesos
  afectados
- Revisados los casos de envío en que devuelve "aceptado con errores"
  para enviarlo como subsanación e incidencia.
- Actualizado roadmap y readme con lo que queda pendiente.
- Arreglado tests que fallaban

JordiBForgeFlow committed:
[REF+IMP] l10n_es_verifactu_oca: verifactu.send.* + mocked tests

- Refactor para enviar las facturas desde verifactu.send.queue y guardar las respuestas en verifactu.send.respose
- Improve tests related with the chaining of invoices
- Mandar incidencia en cabecera cuando los registros se han generado pasados 260s del envio a verifactu
- Add tests with mocked responses from verifactu

almumu committed:
[FIX] l10n_es_verifactu_oca: añadido compatibilidad y herencia que faltaba con aeat.mixin

MeritxellAForgeFlow committed:
[IMP] l10n_es_verifactu_oca: Add verifactu send response

ezerouali commented:
Una pregunta que me acaba de asaltar, son los datos de configuración del módulo, como SIF Name/ID, Developer Name/VAT, etc... que tendríamos que poner ? ... a nosotros mismos ??

anmarmo1 commented:
Hola @ezerouali pues dependerá, hay que firmar una declaración responsable de que el sistema cumple con la normativa, con respecto a que datos hay que poner en las FAQ de AEAT tienes información.
https://sede.agenciatributaria.gob.es/Sede/iva/sistemas-informaticos-facturacion-verifactu/preguntas-frecuentes/certificacion-sistemas-informaticos-declaracion-responsable.html?faqId=524077fe52572910VgnVCM100000dc381e0aRCRD

ezerouali commented:
Hola Anna, y gracias por contestar tan rápido :-) 
Al ser un desarrollo Open Source, y yo solo un implantador, lógicamente y según el reglamento, el SIF no me debería corresponder a mi, de hecho, obtengo lógicamente un error por no estar dado de alta en la AEAT para tal fin.
Solo se nos exige que el software este debidamente certificado por su fabricante, lo que resulta complicado en nuestro supuesto.
Por lo que insisto en la pregunta, pero sospecho que la AEAT no ha previsto nuestra casuística, como por ejemplo:
La Certificación por medio de Declaración responsable, quien la llevaría a cabo ?

jrg-io commented:
@ezerouali en el enlace que te ha pasado @anmarmo1 se puede leer lo siguiente relativo al software de código abierto:

"La empresa que efectúe la programación del código o integre partes de otro software, ya sea o no de código abierto, debe realizar la declaración responsable, incorporar los datos obligatorios e indicar qué componentes utiliza. En el caso de que utilice código abierto, se hará responsable del funcionamiento del mismo, incluyendo los sistemas de seguridad exigidos para preservar la integridad, conservación, accesibilidad, legibilidad, trazabilidad e inalterabilidad de los registros."

Lo que yo interpreto es que si tú comercializas un producto de SIF de código abierto, debes ser tú quien firme la declaración responsable.

anmarmo1 commented:
Según lo que Ignacio comentó, si es código abierto y tú no te puedes hacer responsable de que no hayan modificaciones, el propio cliente puede firmar la declaración responsable, ¿no es así @acysos ?

ljsalvatierra-factorlibre committed:
[IMP] l10n_es_verifactu_oca: new chaining model

ljsalvatierra-factorlibre commented:
Hola @almumu, revisad por favor este nuevo PR donde implemento un nuevo modelo para el encadenamiento como indicáis en el roadmap.

https://github.com/OCA/l10n-spain/pull/4261

almumu commented:
> Hola @almumu, revisad por favor este nuevo PR donde implemento un nuevo modelo para el encadenamiento como indicáis en el roadmap.
> 
> #4261

buenos días @ljsalvatierra-factorlibre, cuando pueda lo reviso, ayer estuvimos viendo algunas cosas que faltan por arreglar en la versión actual y el envío separado de la validación de la factura, que lo está haciendo forgeflow, la idea sería que cuando tengamos eso propongamos que se revise y mergee ya a la OCA. Y posteriormente ya se podría integrar este PR con el encadenamiento a parte.

rafaelbn commented:
¿Sabéis que el Verifactu de Odoo en OpenSource verdad? Lo mismo que el SII.

l10n_es_edi_verifactu

https://github.com/odoo/odoo/pull/197635

Hola @zamberjo , ¿tú sigues trabajando en esto activamente?

anmarmo1 commented:
Hola @rafaelbn, si conocemos la existencia de l10n_es_edi_verifactu pero a parte de no estar en la v.16 creo que no cuadraría con el mapeo de impuestos de OCA.
@zamberjo ya no esta trabajando en esto activamente pero esta @almumu y yo misma, además de un grupo de trabajo que se organizó en el codesprint de Cartagena. En cualquier caso, cualquier aporte es bienvenido

zamberjo commented:
Hola @rafaelbn,
efectivamente, como bien comenta @anmarmo1, ya no estoy trabajando activamente en esto. Fue un arranque inicial en las OCA Days de Tenerife y desde entonces el impulso lo han llevado @almumu y @anmarmo1, que están haciendo un gran trabajo junto con el resto del grupo.

ljsalvatierra-factorlibre committed:
[REF] l10n_es_verifactu_oca: integrate new verifactu invoice model with batch queue sending

almumu committed:
[REF+FIX+IMP] l10n_es_verifactu_oca: Models unification + fixes + imps

- Unificado modelos de verifactu.send.queue con el
  verifactu.invoice.entry que es el encadenamiento
- Refactor y arreglo de test.
- Añadido también algunas mejoras en cuanto a comprobaciones y
  trazabilidad.
- Arreglos y mejoras en el código según revisión del PR.

ljsalvatierra-factorlibre commented on l10n_es_verifactu/data/ir_config_parameter.xml:
```diff
@@ -0,0 +1,9 @@
+<!-- Copyright 2024 Almudena de La Puente <almudena@aurestic.es>
+     License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl). -->
+<odoo noupdate="1">
+    <record id="verifactu_batch" model="ir.config_parameter">
```
El método `_get_verifactu_batch` que accede a esta configuración no se utiliza por lo que veo.

almumu replied:
ahora ya se usa :)

ljsalvatierra-factorlibre commented on l10n_es_verifactu/models/verifactu_mixin.py:
```diff
@@ -0,0 +1,451 @@
+        )
+
+    @api.model
+    def _get_verifactu_batch(self):
```
No se utiliza por lo que veo.

ljsalvatierra-factorlibre commented on l10n_es_verifactu/models/account_fiscal_position.py:
```diff
@@ -0,0 +1,43 @@
+        copy=False,
+        default=True,
+        help="Enable Verifactu for this fiscal position?",
+    )
```
Creo que no se usa.

ljsalvatierra-factorlibre commented on l10n_es_verifactu/models/account_fiscal_position_template.py:
```diff
@@ -0,0 +1,22 @@
+        copy=False,
+        default=True,
+        help="Enable Verifactu for this fiscal position?",
+    )
```
Creo que no se usa.

almumu replied:
quitado

ljsalvatierra-factorlibre commented on l10n_es_verifactu/models/account_journal.py:
```diff
@@ -0,0 +1,92 @@
+    # para verifactu, por lo que este código no nos terminaría de valer.
+    # De momento lo dejamos hasta saber cómo vamos a controlar
+    # el tema de la factura anterior enviada a verifactu para el cálculo
+    # del hash, y el control de modificaciones en las facturas ya enviadas.
```
Entiendo que esto sobra ya.

ljsalvatierra-factorlibre commented on l10n_es_verifactu/models/account_journal.py:
```diff
@@ -0,0 +1,92 @@
+    # De momento lo dejamos hasta saber cómo vamos a controlar
+    # el tema de la factura anterior enviada a verifactu para el cálculo
+    # del hash, y el control de modificaciones en las facturas ya enviadas.
+    restrict_mode_hash_table = fields.Boolean(
+        compute="_compute_restrict_mode_hash_table",
+        store=True,
+        readonly=False,
+    )
+
+    restrict_mode_hash_table_readonly = fields.Boolean(
+        store=True,
+        compute="_compute_restrict_mode_hash_table",
+    )
```
No entiendo el uso de estos campos.

almumu replied:
esto es porque se usa el bloqueo de los diarios de odoo, pero por defecto esos campos son un boolean sin más, aquí se transforma en un compute para que si tienen activado verifactu y son asientos de venta no permita modificar las facturas

ljsalvatierra-factorlibre replied:
Gracias!

ljsalvatierra-factorlibre commented on l10n_es_verifactu/models/account_move.py:
```diff
@@ -0,0 +1,679 @@
+                invoice.verifactu_enabled = (
+                    invoice.fiscal_position_id
+                    and invoice.fiscal_position_id.aeat_active
+                ) or not invoice.fiscal_position_id
```
Quizás deberíamos avisar al usuario de que la posición fiscal establecida requiere `aeat_active`. El modelo `account.fiscal.position` y su modelo plantilla contienen el campo `verifactu_enabled` relacionado sobre `company_id`. Me parece confuso la gestión actual.

almumu replied:
cambiado

ljsalvatierra-factorlibre commented on l10n_es_verifactu/models/account_move.py:
```diff
@@ -0,0 +1,679 @@
+        return self.verifactu_description or self.company_id.verifactu_description
+
+    def _get_document_date(self):
+        """
+        TODO: this method is the same in l10n_es_aeat_sii_oca, so I think that
+        it should be directly in l10n_es_aeat
+        """
+        return self.invoice_date
+
+    def _aeat_get_partner(self):
+        """
+        TODO: this method is the same in l10n_es_aeat_sii_oca, so I think that
+        it should be directly in l10n_es_aeat
+        """
+        return self.commercial_partner_id
+
+    def _get_document_fiscal_date(self):
+        """
+        TODO: this method is the same in l10n_es_aeat_sii_oca, so I think that
+        it should be directly in l10n_es_aeat
+        """
+        return self.invoice_date
+
+    def _get_mapping_key(self):
+        """
+        TODO: this method is the same in l10n_es_aeat_sii_oca, so I think that
+        it should be directly in l10n_es_aeat
+        """
+        return self.move_type
```
TODO?

almumu replied:
esto está en el roadmap, hay que revisar qué métodos se pueden unificar y mover al l10n_es_aeat

ljsalvatierra-factorlibre commented on l10n_es_verifactu/models/account_move.py:
```diff
@@ -0,0 +1,679 @@
+        serial_number = (self.name or "")[0:60]
+        if self.thirdparty_invoice:
+            serial_number = self.thirdparty_number[0:60]
+        return serial_number
```
TODO?

almumu replied:
lo mismo

ljsalvatierra-factorlibre commented on l10n_es_verifactu/models/account_move.py:
```diff
@@ -0,0 +1,679 @@
+            f"ImporteTotal={amountTotal}&"
+            f"Huella={previousHash}&"
+            f"FechaHoraHusoGenRegistro={registrationDate}"
+        )
```
camel?

almumu replied:
cambiado

ljsalvatierra-factorlibre commented on l10n_es_verifactu/models/account_move.py:
```diff
@@ -0,0 +1,679 @@
+                            "IDEmisorFactura": company_vat,
+                            "NumSerieFactura": orig_serial_number,
+                            "FechaExpedicionFactura": orig_document_date,
+                        }
+                    }
+                    inv_dict["FacturasRectificadas"].append(origin_data)
+                # inv_dict["ImporteRectificacion"] = {
+                #     "BaseRectificada": abs(origin.amount_untaxed_signed),
+                #     "CuotaRectificada": abs(
+                #         origin.amount_total_signed - origin.amount_untaxed_signed
+                #     ),
+                # }
```
Indicar en README si no va a ser implementado de momento.

almumu replied:
está en el readme como "Casos no cubiertos"

ljsalvatierra-factorlibre commented on l10n_es_verifactu/models/account_move.py:
```diff
@@ -0,0 +1,679 @@
+                    "CalificacionOperacion": operation_type,
+                }
+                # si es exenta:
+                # "OperacionExenta": "", # TODO
```
Indicar en ROADMAP

almumu replied:
puesto

ljsalvatierra-factorlibre commented on l10n_es_verifactu/models/account_move.py:
```diff
@@ -0,0 +1,679 @@
+        return res
+
+    def _check_verifactu_configuration(self):
+        if not self.fiscal_position_id:
+            raise UserError(
+                _(
+                    "The invoice %s cannot be sent to Verifactu because it "
+                    "does not have a fiscal position."
+                )
+                % self.name
+            )
```
Al calcular `verifactu_enabled` permitimos que no tenga posición fiscal.

almumu replied:
arreglado

ljsalvatierra-factorlibre commented on l10n_es_verifactu/models/account_move.py:
```diff
@@ -0,0 +1,679 @@
+
+    def _get_verifactu_chain_context(self):
+        """Use a unique chain for each company for standard invoicing."""
+        return ("res.company", self.company_id)
```
No se usa.

almumu replied:
quitado

ljsalvatierra-factorlibre commented on l10n_es_verifactu/models/account_move.py:
```diff
@@ -0,0 +1,679 @@
+        return ("res.company", self.company_id)
+
+    def cancel_verifactu(self):
+        raise NotImplementedError
```
Indicar en ROADMAP.

almumu replied:
está en Casos no cubiertos

ljsalvatierra-factorlibre commented on l10n_es_verifactu/models/account_move.py:
```diff
@@ -0,0 +1,679 @@
+        for rec in self:
+            if (
+                rec.last_verifactu_invoice_entry_id
+                and not rec.last_verifactu_invoice_entry_id.send_state == "not_sent"
```
`rec.last_verifactu_invoice_entry_id.send_state != "not_sent"`

Deberíamos comprobar aquí `aeat_state`, no solo en la vista.

almumu replied:
hecho

ljsalvatierra-factorlibre commented on l10n_es_verifactu/models/res_company.py:
```diff
@@ -0,0 +1,45 @@
+                    )
+                    if journals:
+                        journals.write({"verifactu_enabled": True})
+        return res
```
Quizás debería ser un campo calculado en el diario?

almumu replied:
no, esto se hizo en los oca days, si se activa verifactu en la compañía, todos los diarios de venta pasan a ser verifactu_enabled=True, pero debe permitir luego desactivar un diario concreto si se quisiera. Se hizo para poder excluir diarios

ljsalvatierra-factorlibre commented on l10n_es_verifactu/models/verifactu_invoice_entry.py:
```diff
@@ -0,0 +1,369 @@
+    _order = "id desc"
+    _rec_name = "document_hash"
+
+    verifactu_chaining_id = fields.Many2one(
```
Campo requerido.

almumu replied:
ok

ljsalvatierra-factorlibre commented on l10n_es_verifactu/models/verifactu_invoice_entry.py:
```diff
@@ -0,0 +1,369 @@
+        ondelete="restrict",
+    )
+    model = fields.Char(readonly=True)
+    document_id = fields.Many2oneReference(
```
Campo requerido.

almumu replied:
hecho

ljsalvatierra-factorlibre commented on l10n_es_verifactu/models/verifactu_invoice_entry.py:
```diff
@@ -0,0 +1,369 @@
+        required=True,
+        readonly=True,
+    )
+    aeat_json_data = fields.Text(
```
Campo requerido.

almumu replied:
hecho

ljsalvatierra-factorlibre commented on l10n_es_verifactu/models/verifactu_invoice_entry.py:
```diff
@@ -0,0 +1,369 @@
+        string="Chaining",
+        ondelete="restrict",
+    )
+    model = fields.Char(readonly=True)
```
Campo requerido.

almumu replied:
hecho

ljsalvatierra-factorlibre commented on l10n_es_verifactu/models/verifactu_invoice_entry.py:
```diff
@@ -0,0 +1,369 @@
+
+    @api.model
+    def _cron_send_documents_to_verifactu(self):
+        for company in self.env["res.company"].search(
+            [("verifactu_enabled", "=", True)]
+        ):
```
Quizás deberíamos hacerlo por cadena en vez de por `res.company`.

almumu replied:
hecho

ljsalvatierra-factorlibre commented on l10n_es_verifactu/models/verifactu_invoice_entry.py:
```diff
@@ -0,0 +1,369 @@
+            cancellation
+        :return Dict with header data depending on cancellation
+        """
+        # todo: implementar RemisionVoluntaria
```
Ya lo tenemos no?

almumu replied:
si

ljsalvatierra-factorlibre commented on l10n_es_verifactu/models/verifactu_invoice_entry.py:
```diff
@@ -0,0 +1,369 @@
+                str(linea["CodigoErrorRegistro"]),
+                str(linea["DescripcionErrorRegistro"]),
+            )
+            if linea["CodigoErrorRegistro"] == 3000:
```
Meter en el ROADMAP el campo de tipo selección para los códigos.

almumu replied:
hecho

ljsalvatierra-factorlibre commented on l10n_es_verifactu/models/verifactu_invoice_entry.py:
```diff
@@ -0,0 +1,369 @@
+        doc_vals["aeat_send_error"] = send_error
+        return doc_vals
+
+    def _send_documents_to_verifactu(self):
```
Refactorizar de este método para dividirlo en funciones.

almumu replied:
hecho

ljsalvatierra-factorlibre commented on l10n_es_verifactu/models/verifactu_mixin.py:
```diff
@@ -0,0 +1,451 @@
+    qrcode = None
+    _logger.error(err)
+
+###########################################
+# revisar los imports que no hagan falta
+# cuando funcione bien el _connect_aeat sin tener que poner
+# el forbid_entities, y se pueda borrar la función _connect_verifactu
+# de este fichero para usar la del aeat_mixin
```
Quitamos?

almumu replied:
hecho

ljsalvatierra-factorlibre commented on l10n_es_verifactu/models/verifactu_mixin.py:
```diff
@@ -0,0 +1,451 @@
+        "greater o equal to 100 000 000,00 euros.",
+        compute="_compute_verifactu_macrodata",
+    )
+    verifactu_csv = fields.Char(copy=False, readonly=True)
```
No consigo ver donde se establece.

almumu replied:
se establece en _process_response_line_doc_vals

ljsalvatierra-factorlibre replied:
visto, gracias! Habia leído la linea, pero no me había fijado en `document`

`verifactu_invoice_entry.document.write(doc_vals)`

ljsalvatierra-factorlibre commented on l10n_es_verifactu/models/verifactu_invoice_entry.py:
```diff
@@ -0,0 +1,369 @@
+        respuestaLineas = "RespuestaLinea" in res and res["RespuestaLinea"] or []
+        for linea in respuestaLineas:
+            invoice_num = linea["IDFactura"]["NumSerieFactura"]
+            document_models = self.env[
+                "verifactu.mixin"
+            ]._selection_verifactu_reference_models()
```
No llamar a nivel de linea dentro del bucle.

almumu replied:
hecho

ljsalvatierra-factorlibre commented on l10n_es_verifactu/models/verifactu_mixin.py:
```diff
@@ -0,0 +1,451 @@
+        self.ensure_one()
+
+        # Always use company for invoice chaining
+        chaining = self.company_id.verifactu_chaining_id
```
Habíamos comentado de poder utilizar múltiples cadenas para un mismo `res.company`.

almumu replied:
he creado un método _get_chaining en verifactu.mixin para obtener el encadenamiento, en el account.move lo coge de la compañía, para el pos se podrá obtener del pos config o de la compañía

ljsalvatierra-factorlibre commented on l10n_es_verifactu/readme/ROADMAP.rst:
```diff
@@ -0,0 +1,13 @@
+ * Refactorización SII-Verifactu en l10n_es_aeat cuando estén todos los procesos claros
```
Actualizar con lo ya implementado y temas pendientes.

almumu replied:
hecho

almumu commented:
@ljsalvatierra-factorlibre muchas gracias por la revisión, ya he subido los cambios.

ljsalvatierra-factorlibre commented:
> @ljsalvatierra-factorlibre muchas gracias por la revisión, ya he subido los cambios.

Muchas gracias a ti! Hacemos pruebas funcionales y os comentamos en cuanto podamos.

ljsalvatierra-factorlibre commented:
> @ljsalvatierra-factorlibre muchas gracias por la revisión, ya he subido los cambios.

Hola @almumu , no consigo hacer el envío:

`Fault('Codigo[4104].Error en la cabecera: el valor del campo NIF del bloque ObligadoEmision no está identificado.. NIF:99999910G. NOMBRE_RAZON:Dummy Clothes, S.L.')`

```
(Pdb) header
{'ObligadoEmision': {'NombreRazon': 'Dummy Clothes, S.L.', 'NIF': '99999910G'}, 'RemisionVoluntaria': {'Incidencia': 'S'}}
```

He probado tanto con incidencia como sin incidencia.

anmarmo1 commented:
Hola @ljsalvatierra-factorlibre Los nifs se comprueban. Tanto de la Compañía como del Cliente como del Desarrollador. Han de ser NiF válidos y estar Censados en AEAT, al menos diría que el de la compañía seguro y ha de coincidir con el nif del Certificado. Hacen esas comprobaciones.

ljsalvatierra-factorlibre commented:
> Hola @ljsalvatierra-factorlibre Los nifs se comprueban. Tanto de la Compañía como del Cliente como del Desarrollador. Han de ser NiF válidos y estar Censados en AEAT, al menos diría que el de la compañía seguro y ha de coincidir con el nif del Certificado. Hacen esas comprobaciones.

Intento con otra configuración, gracias.

anmarmo1 commented:
Hola ¿podemos revisar este pr para ver si se mergea? @pedrobaeza @carlosdauden @JordiBForgeFlow @omar7r @acysos @ljsalvatierra-factorlibre @ValentinVinagre @flachicasidoo

pedrobaeza commented:
Disculpad, que está siendo un agosto más movido de lo esperado. Esta semana reviso y fusiono.

ozono committed:
[FIX] l10n_es_verifactu_oca: 2 decimals in URL

Formateo importe a dos decimales en URL QR para cumplir límite AEAT y
evitar errores con números flotantes.

ValentinVinagre commented:
@HaraldPanten FYI

pedrobaeza committed:
[IMP] l10n_es_aeat_verifactu_oca: Final adjustments

- Final renaming to l10n_es_verifactu_oca.
- Homogeneous code styling without empyt line inside methods.
- Code compactation friendly to ruff.
- Code optimization.
- A little of tests reorganization.
- Labelling everywhere with "VERI*FACTU" naming according the official
  naming and case: https://sede.agenciatributaria.gob.es/Sede/iva/sistemas-informaticos-facturacion-verifactu.html
- FIX: Send IndicadorMultiplesOT = S only when multiple companies have
  enabled VeriFactu, not when they have Spain as country.
- Proper models naming (without plurals) + descriptions.
- Field names and form labels rationalized with the following criteria:
  * For field technical names, all are prefixed with `verifactu_`.
  * For field names (labels), all are prefixed with `VERI*FACTU ` for
    being all grouped when doing a filter.
  * For form labels, the prefix is explicitly removed for not
    overloading the UI and not repeating information.
- Rename verifactu.invoice.entry.response~datetime to date_response for
  not collapsing with datetime library.
- Rellocate some fields of account.move into verifactu.mixing for being
  available to all the VERI*FACTU documents.
- Fiscal position VERI*FACTU settings moved to a page in the notebook
  for better visual ergonomy.
- Invoice VERI*FACTU page layout restructured for better ergonomy.
- Show all the VERI*FACTU checks errors at once and without repeating
  strings, with a huge refactorization and also contemplating other
  possible tax agencies.
- Remove check about requiring a fiscal position in the invoice. The
  usual case for national customers is to not set "Régimen Nacional", so
  now forcing to set it will increase the entry barrier.
- Translations to Spanish.

pedrobaeza commented:
@almumu estoy ya terminando la revisión y retoque del PR. Como en esta rama no tengo permiso para escribir, ¿abro nuevo PR, o te digo y sobreescribes tu rama con la mía (hay reordenamiento/squash de commits)?

anmarmo1 commented:
Hola @pedrobaeza @almumu va a estar fuera unos días hasta el 2, no he entendido muy bien lo de los PR's pero como veas para no tener que esperar una semana, si no la esperamos. Un saludo

pedrobaeza commented:
Gracias por contestar, Anna. Bueno, voy a seguir haciendo los cambios y cuando termine, lo vemos.

pedrobaeza commented:
Acabo de subir los que en principio deberían ser los últimos cambios. Lo primero, debido a que alguien ha subido ya un módulo a apps.odoo.com con el nombre `l10n_es_verifactu`, lo hemos renombrado finalmente a `l10n_es_verifactu_oca`.

Después, el historial de commits ha sido racionalizado para tener mensajes de commits homogéneos, y juntando commits de trabajos intermedios para intentar agruparlos por conjuntos, pero siempre respetando la atribución.

Por último, os paso lo escrito en el mensaje de commit (en inglés) de esos últimos cambios:

- Final renaming to l10n_es_verifactu_oca.
- Homogeneous code styling without empyt line inside methods.
- Code compactation friendly to ruff.
- Code optimization.
- A little of tests reorganization.
- Labelling everywhere with "VERI*FACTU" naming according the official naming and case: https://sede.agenciatributaria.gob.es/Sede/iva/sistemas-informaticos-facturacion-verifactu.html
- FIX: Send IndicadorMultiplesOT = S only when multiple companies have enabled VeriFactu, not when they have Spain as country.
- Proper models naming (without plurals) + descriptions.
- Field names and form labels rationalized with the following criteria:
  * For field technical names, all are prefixed with `verifactu_`.
  * For field names (labels), all are prefixed with `VERI*FACTU ` for
    being all grouped when doing a filter.
  * For form labels, the prefix is explicitly removed for not
    overloading the UI and not repeating information.
- Rename verifactu.invoice.entry.response~datetime to date_response for not collapsing with datetime library.
- Rellocate some fields of account.move into verifactu.mixin for being available to all the VERI*FACTU documents.
- Fiscal position VERI*FACTU settings moved to a page in the notebook for better visual ergonomy.
- Invoice VERI*FACTU page layout restructured for better ergonomy.
- Show all the VERI*FACTU checks errors at once and without repeating strings, with a huge refactorization and also contemplating other possible tax agencies.
- Remove check about requiring a fiscal position in the invoice. The usual case for national customers is to not set "Régimen Nacional", so now forcing to set it will increase the entry barrier.
- Translations to Spanish.

@aritzolea @ljsalvatierra-factorlibre pusistéis "Requested changes" hace tiempo. ¿Podéis revisar para quitarlo o comentarnos si veis algo que falta? Veo que también habéis hecho el PR del PoS, aunque ahora después de estos cambios habrá que hacer algunas adaptaciones. Os puedo ayudar si queréis.

Vamos a intentar dejar esto fusionado finalmente esta semana, ¿os parece?

aritzolea commented:
> Acabo de subir los que en principio deberían ser los últimos cambios. Lo primero, debido a que alguien ha subido ya un módulo a apps.odoo.com con el nombre `l10n_es_verifactu`, lo hemos renombrado finalmente a `l10n_es_verifactu_oca`.
> 
> Después, el historial de commits ha sido racionalizado para tener mensajes de commits homogéneos, y juntando commits de trabajos intermedios para intentar agruparlos por conjuntos, pero siempre respetando la atribución.
> 
> Por último, os paso lo escrito en el mensaje de commit (en inglés) de esos últimos cambios:
> 
> * Final renaming to l10n_es_verifactu_oca.
> * Homogeneous code styling without empyt line inside methods.
> * Code compactation friendly to ruff.
> * Code optimization.
> * A little of tests reorganization.
> * Labelling everywhere with "VERI*FACTU" naming according the official naming and case: https://sede.agenciatributaria.gob.es/Sede/iva/sistemas-informaticos-facturacion-verifactu.html
> * FIX: Send IndicadorMultiplesOT = S only when multiple companies have enabled VeriFactu, not when they have Spain as country.
> * Proper models naming (without plurals) + descriptions.
> * Field names and form labels rationalized with the following criteria:
>   
>   * For field technical names, all are prefixed with `verifactu_`.
>   * For field names (labels), all are prefixed with `VERI*FACTU ` for
>     being all grouped when doing a filter.
>   * For form labels, the prefix is explicitly removed for not
>     overloading the UI and not repeating information.
> * Rename verifactu.invoice.entry.response~datetime to date_response for not collapsing with datetime library.
> * Rellocate some fields of account.move into verifactu.mixin for being available to all the VERI*FACTU documents.
> * Fiscal position VERI*FACTU settings moved to a page in the notebook for better visual ergonomy.
> * Invoice VERI*FACTU page layout restructured for better ergonomy.
> * Show all the VERI*FACTU checks errors at once and without repeating strings, with a huge refactorization and also contemplating other possible tax agencies.
> * Remove check about requiring a fiscal position in the invoice. The usual case for national customers is to not set "Régimen Nacional", so now forcing to set it will increase the entry barrier.
> * Translations to Spanish.
> 
> @aritzolea @ljsalvatierra-factorlibre pusistéis "Requested changes" hace tiempo. ¿Podéis revisar para quitarlo o comentarnos si veis algo que falta? Veo que también habéis hecho el PR del PoS, aunque ahora después de estos cambios habrá que hacer algunas adaptaciones. Os puedo ayudar si queréis.
> 
> Vamos a intentar dejar esto fusionado finalmente esta semana, ¿os parece?

No veo la opción de eliminar los cambios solicitados, por que están pisados por otros commits, me aparece como readonly. Aun así ya son cambios antiguos que están mas que cubiertos, adelante por nuestra parte. Gracias!

pedrobaeza commented:
Gracias a todos los que habéis intervenido. Procedo a fusionar.

/ocabot merge nobump

ljsalvatierra-factorlibre commented:
Hola @almumu, para comentar los próximos desarrollos del roadmap? Por ejemplo, si se va a abordar:

* Envío separado de la confirmación de la factura (cron.trigger/queque.job)

almumu commented:
hola @ljsalvatierra-factorlibre , tengo entendido que los nuevos desarrollos que se hagan a partir de ahora sobre verifactu se tratarán como issues para no duplicar tareas y esfuerzos. De momento de ese punto en concreto creo que no se ha comentado nada. El envío actualmente ya se hace separado de la confirmación de la factura, se ejecuta el cron cada minuto enviando todo lo que quede pendiente de comunicar, faltaba ver si había una forma más óptima de implementarlo mediante cron.triggers o queue.job, pero siempre teniendo en cuenta que (salvo que haya habido algún cambio) el tiempo máximo para comunicar la factura a hacienda una vez emitida es de 120 segundos. Y tiene que haber un tiempo de espera mínimo de 60 segundos entre envíos.

ljsalvatierra-factorlibre commented:
> hola @ljsalvatierra-factorlibre , tengo entendido que los nuevos desarrollos que se hagan a partir de ahora sobre verifactu se tratarán como issues para no duplicar tareas y esfuerzos. De momento de ese punto en concreto creo que no se ha comentado nada. El envío actualmente ya se hace separado de la confirmación de la factura, se ejecuta el cron cada minuto enviando todo lo que quede pendiente de comunicar, faltaba ver si había una forma más óptima de implementarlo mediante cron.triggers o queue.job, pero siempre teniendo en cuenta que (salvo que haya habido algún cambio) el tiempo máximo para comunicar la factura a hacienda una vez emitida es de 120 segundos. Y tiene que haber un tiempo de espera mínimo de 60 segundos entre envíos.

Gracias @almumu, continuamos entonces con las issues. Sobre los trabajos en cola, entiendo que la solución actual es la correcta, con la acción planificada.
